wp/lib/bap_wp/src/constraint.ml

```diff
@@ -233,51 +233,48 @@ let get_refuted_goals ?filter_out:(filter_out = []) (constr : t)
+      then Seq.singleton
             { goal = { g with goal_val = goal_val };
               path = current_path;
               reg_map = current_registers }
-        | Solver.UNKNOWN ->
-          failwith (Format.sprintf "get_refuted_goals: Unable to resolve %s" g.goal_name)
-      end
+      else
+        failwith (Format.sprintf "get_refuted_goals: Unable to resolve %s" g.goal_name)
     | ITE (jmp, cond, c1, c2) ->
       let cond_val = Expr.substitute cond olds news in
       let cond_res = eval_model_exn model cond_val in
       let current_registers = update_current_regs
           model olds news jmp current_registers filter_out in
-      begin
-        match Solver.check solver [cond_res] with
-        | Solver.SATISFIABLE ->
-          let current_path = Jmp.Map.set current_path ~key:jmp ~data:true in
-          worker c1 current_path current_registers olds news
-        | Solver.UNSATISFIABLE ->
-          let current_path = Jmp.Map.set current_path ~key:jmp ~data:false in
-          worker c2 current_path current_registers olds news
-        | Solver.UNKNOWN ->
-          failwith (Format.sprintf "get_refuted_goals: Unable to resolve branch \
-                                    condition at %s" (jmp |> Term.tid |> Tid.to_string))
-      end
+      if Z3.Boolean.is_true cond_res
+      then
+        let current_path = Jmp.Map.set current_path ~key:jmp ~data:true in
+        worker c1 current_path current_registers olds news
+      else if Z3.Boolean.is_false cond_res
+      then
+        let current_path = Jmp.Map.set current_path ~key:jmp ~data:false in
+        worker c2 current_path current_registers olds news
+      else       
+        failwith (Format.sprintf "get_refuted_goals: Unable to resolve branch \
+                                    condition %s at %s" 
+                                    (Z3.Expr.to_string cond_res)
+                                    (jmp |> Term.tid |> Tid.to_string))
     | Clause (hyps, concs) ->
       let hyps_false =
         let hyp_vals =
           List.map hyps ~f:(fun h -> eval_model_exn model (eval_aux h olds news ctx)) in
-        match Solver.check solver hyp_vals with
-        | Solver.SATISFIABLE -> false
-        | Solver.UNSATISFIABLE -> true
-        | Solver.UNKNOWN ->
-          failwith "get_refuted_goals: Unable to resolve value of hypothesis"
+          if (List.for_all hyp_vals ~f:(fun e -> Z3.Boolean.is_true e || Z3.Boolean.is_false e))
+          then not (List.for_all hyp_vals ~f:Z3.Boolean.is_true)
+          else failwith "Unexpected Expression processing Clause"
       in
       if hyps_false then
         Seq.empty
       else
         List.fold concs ~init:Seq.empty ~f:(fun accum c ->
             Seq.append (worker c current_path current_registers olds news) accum)
     | Subst (e, o, n) ->
-      let n' = List.map n ~f:(fun x -> Expr.substitute x olds news) in
+      let n' = List.map n ~f:(fun x -> Expr.substitute x olds news |> eval_model_exn model) in
```

Comment:
What does this enable?

Reply:
I guess it's related to your above comment. But still walking through the state of these terms before and after the call to `eval_model` would be useful.